cluster-provision/centos9/scripts/vm.sh

```diff
@@ -184,15 +192,100 @@ if [ "${NUMA}" -gt 1 ]; then
-  -device intel-iommu,intremap=on,caching-mode=on -device intel-hda -device hda-duplex -device AC97 \
-  -uuid $(cat /proc/sys/kernel/random/uuid) \
-  ${QEMU_ARGS}
+if [ "$(uname -m)" == "s390x" ]; then
+  # As per https://www.qemu.org/docs/master/system/s390x/bootdevices.html#booting-without-bootindex-parameter -drive if=virtio can't be specified with bootindex for s390x
+  qemu_system_cmd="qemu-system-s390x \
+    -enable-kvm \
+    -drive format=qcow2,file=${next},if=none,cache=unsafe,id=drive1 ${block_dev_arg} \
+    -device virtio-blk,drive=drive1,bootindex=1 \
+    -device virtio-net-ccw,netdev=network0,mac=52:55:00:d1:55:${n} \
+    -netdev tap,id=network0,ifname=tap${n},script=no,downscript=no \
+    -device virtio-rng \
+    -initrd /initrd.img \
+    -kernel /vmlinuz \
+    -append \"$(cat /kernel.s390x.args) $(cat /additional.kernel.args) ${KERNEL_ARGS}\" \
+    -vnc :${n} \
+    -cpu host \
+    -m ${MEMORY} \
+    -smp ${CPU} ${numa_arg} \
+    -serial pty \
+    -machine s390-ccw-virtio,accel=kvm \
+    -uuid $(cat /proc/sys/kernel/random/uuid) \
+    ${QEMU_ARGS}"
+else
+  #Docs: https://www.qemu.org/docs/master/system/invocation.html
+  qemu_system_cmd="qemu-system-x86_64 \
+    -enable-kvm \
+    -drive format=qcow2,file=${next},if=virtio,cache=unsafe ${block_dev_arg} \
+    -device virtio-net-pci,netdev=network0,mac=52:55:00:d1:55:${n} \
+    -netdev tap,id=network0,ifname=tap${n},script=no,downscript=no \
+    -device virtio-rng-pci \
+    -initrd /initrd.img \
+    -kernel /vmlinuz \
+    -append \"$(cat /kernel.args) $(cat /additional.kernel.args) ${KERNEL_ARGS}\" \
+    -vnc :${n} \
+    -cpu host,migratable=no,+invtsc \
+    -m ${MEMORY} \
+    -smp ${CPU} ${numa_arg} \
+    -serial pty \
+    -machine q35,accel=kvm,kernel_irqchip=split \
+    -device intel-iommu,intremap=on,caching-mode=on \
+    -device intel-hda \
+    -device hda-duplex \
+    -device AC97 \
+    -uuid $(cat /proc/sys/kernel/random/uuid) \
+    ${QEMU_ARGS}"
+fi
+
+# Remove secondary network devices from qemu_system_cmd and move them to qemu_monitor_cmds, so
+# that those devices are later added after VM is started using qemu monitor to avoid
+# primary network interface to be named other than eth0. This is mainly required for s390x, as 
+# otherwise if primary interface is other than eth0, it is not getting the IP from dhcp server.
+qemu_monitor_cmds=()
+IFS=' ' read -r -a qemu_parts <<< "$qemu_system_cmd"
+for ((i = 0; i < ${#qemu_parts[@]}; i++)); do
+  part="${qemu_parts[$i]}"
+  nxtpart="${qemu_parts[$i + 1]}"
+  # Check for secondary network devices and move them to qemu_monitor_cmds
+  if { [ "$part" == "-netdev" ] && [[ "$nxtpart" == *"secondarynet"* ]]; } || \
+     { [ "$part" == "-device" ] && [[ "$nxtpart" == *"virtio-net-ccw"* ]] && [[ "$nxtpart" == *"secondarynet"* ]]; }; then
+    qemu_system_cmd=$(echo "$qemu_system_cmd" | sed "s/ -$part $nxtpart//")
+    qemu_monitor_cmds+=("${part}_add $nxtpart")
+  fi
+done
```

Comment:
This feels strange - we first create the `qemu_system_cmd` and then we remove the secondary network devices? Could we just not include them when creating `qemu_system_cmd` ?